content/blog/2024-08-29-valkey-memory-efficiency-8-0.md

```diff
@@ -0,0 +1,148 @@
+
+1. Modify value for a given index (slot) - If a key gets added or removed, increment or decrement the value for that given slot by 1 respectively.
+2. Cumulative frequency until each index (slot) - For a given number representing a key between 1 and the total number of keys, return the index (slot) which covers the particular key.
+
+If approached naively, the former and latter operation would take O(1) and O(N) respectively. However, we want to minimize the latter operation’s time complexity and avoid regression in the operations stated above. Hence, a [binary indexed tree (BIT) or fenwick tree](https://www.topcoder.com/thrive/articles/Binary%20Indexed%20Trees) which provides the above functionality with a minimal memory overhead (~1 MB per node) and the time complexity is also bounded to O(M log N) for both the operations where M = number of modification(s) and N = number of slots. This enables skipping over empty slot efficiently while iterating over the keyspace as well as finding a slot for a given key index in logarithmic time via binary search over the cummulative sum maintained by the BIT. 
+
+Overall with this new approach, the benefits are: 
+
+1. Removes additional memory overhead in cluster mode: Get rid of two pointers (16 bytes) per key to keep the mapping of slot to keys.
+2. Rehashing operation is CPU intensive: By default, limited number of buckets are allocated in a dictionary and it expands/shrinks dyanmically based on usage. While undergoing rehashing, all the data needs to be moved from an old dictionary to a new dictionary. With Valkey 7.2, a global dictionary being shared across all the slots, all the keys gets stored under a single dictionary and each time the fill factor (number of keys / number of buckets) goes above 1.5, the dictionary needs to a large dictionary (multiple of 2) and move large amount of keys. As this operation is performed on the fly, it causes increase in latency for regular command operations while it’s ongoing. With this optimization, the impact of rehashing is localized to the specific dictionary undergoing the process and only subset of key needs to be moved. 
+
+### Optimization 2 - [Key embedding into dictionary entry](https://github.com/valkey-io/valkey/pull/541)
+
+After the dictionary per slot change, the memory layout of dictionary entry in cluster mode is the following, there are three pointers (key, value, and next).  The key pointer points to a SDS([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md)) which contains the actual key data. As key is immutable, without bringing in much complexity, it can be embedded into the dictionary entry which has the same lifetime as the former.
```

Comment:
` SDS([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md))` -> ` SDS ([simple dynamic string](https://github.com/antirez/sds/blob/master/README.md))` (needs a space next to the open paren)